### PR TITLE
Allow import via python datatable.

### PR DESCRIPTION
**Pull request by pseudotensor**
Datatable is a multi-threaded data ingestion code whose API is used to read into a DMatrix.  While the DMatrix creation is sometimes faster than numpy, the key ability is to use datatable to read csv or other files, which can be 10X or more faster than pandas.

Datatable also memory maps to disk each column, saving lots of memory and speed compared to pandas or even numpy.

https://github.com/h2oai/datatable

### CODE REVIEW

**Comment by pseudotensor** [comment]
@RAMitchell

**Comment by RAMitchell** [comment]
Yes you should provide a google test for coverage. Also I recommend using git commit amend instead of pushing multiple commits for minor changes.

**Comment by pseudotensor** [comment]
There is no actual c api for python datatable

**Comment by pseudotensor** [comment]
Ok.  Does the python test I added not get counted?  If not, why not?

**Comment by pseudotensor** [comment]
Also, relatedly, how do i ensure test will have pip installed datatable??  requirements in setup.py show only scipy and numpy.  How is pandas installed?

**Comment by RAMitchell** [comment]
The code coverage works by compiling the c++ code and tests with a special flag and seeing which code paths are used/not used. So python is not included. 

Not sure about pandas. You will have to add a pip install step to this file https://github.com/dmlc/xgboost/blob/master/tests/travis/setup.sh.

**Comment by RAMitchell** [comment]
@pseudotensor had a closer look at this today. I get the following output when running the benchmark:
```
Generating dataset: 1000000 rows * 50 columns
0.25/0.75 test/train split
Number of Nans: 5000637
Generate Time: 3.525233268737793 seconds
dt prepare1 Time: 0.570749044418335 seconds
np->DMatrix Start
np->DMatrix1 Time: 0.16157031059265137 seconds
np->DMatrix2 Time: 0.045729875564575195 seconds
Training with 'gpu_exact'
[19:38:55] Allocated 1683MB on [0] TITAN X (Pascal), 10345MB remaining.
[19:38:55] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[0]     test-error:0.056188
[19:38:56] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[19:38:56] Allocated 87MB on [0] TITAN X (Pascal), 10245MB remaining.
[19:38:56] Allocated 263MB on [0] TITAN X (Pascal), 9977MB remaining.
[1]     test-error:0.055752
[19:38:56] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[2]     test-error:0.05392
[19:38:56] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[3]     test-error:0.053116
[19:38:56] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[4]     test-error:0.052728
Train Time: 2.998378276824951 seconds
dt prepare2 Time: 0.00042366981506347656 seconds
dt->DMatrix Start
dt->DMatrix1 Time: 0.3213200569152832 seconds
dt->DMatrix2 Time: 0.09867191314697266 seconds
Training with 'gpu_exact'
[19:38:58] Allocated 1683MB on [0] TITAN X (Pascal), 10345MB remaining.
[19:38:58] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[0]     test-error:0.056188
[19:38:59] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[19:38:59] Allocated 87MB on [0] TITAN X (Pascal), 10245MB remaining.
[19:38:59] Allocated 263MB on [0] TITAN X (Pascal), 9977MB remaining.
[1]     test-error:0.055752
[19:38:59] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[2]     test-error:0.05392
[19:38:59] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[3]     test-error:0.053116
[19:38:59] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[4]     test-error:0.052728
Train Time: 2.530510902404785 seconds
dt prepare2 Time: 0.00038242340087890625 seconds
dt->DMatrix Start
dt->DMatrix1 Time: 0.22930073738098145 seconds
dt->DMatrix2 Time: 0.07219409942626953 seconds
Training with 'gpu_exact'
[19:39:01] Allocated 1683MB on [0] TITAN X (Pascal), 10345MB remaining.
[19:39:01] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[0]     test-error:0.056188
[19:39:01] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[19:39:01] Allocated 87MB on [0] TITAN X (Pascal), 10245MB remaining.
[19:39:01] Allocated 263MB on [0] TITAN X (Pascal), 9977MB remaining.
[1]     test-error:0.055752
[19:39:02] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[2]     test-error:0.05392
[19:39:02] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[3]     test-error:0.053116
[19:39:02] /research/ram41/xgboost-build-helper/ubuntu/xgboost/src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 126 extra nodes, 0 pruned nodes, max_depth=6
[4]     test-error:0.052728
Train Time: 2.540203332901001 seconds
Generating dataset: 1000000 rows * 50 columns
0.25/0.75 test/train split
Number of Nans: 4999969
Generate Time: 3.4246304035186768 seconds
dt prepare1 Time: 0.512869119644165 seconds
np->DMatrix Start
np->DMatrix1 Time: 0.17965316772460938 seconds
np->DMatrix2 Time: 0.05552244186401367 seconds
Training with 'gpu_hist'
[19:39:07] Allocated 82MB on [0] TITAN X (Pascal), 11939MB remaining.
[19:39:07] Allocated 24MB on [0] TITAN X (Pascal), 11913MB remaining.
[0]     test-error:0.056312
[19:39:07] Allocated 87MB on [0] TITAN X (Pascal), 11809MB remaining.
[1]     test-error:0.055784
[2]     test-error:0.055256
[3]     test-error:0.053432
[4]     test-error:0.053016
Train Time: 1.2012369632720947 seconds
dt prepare2 Time: 0.0003352165222167969 seconds
dt->DMatrix Start
dt->DMatrix1 Time: 0.381300687789917 seconds
dt->DMatrix2 Time: 0.11515641212463379 seconds
Training with 'gpu_hist'
[19:39:09] Allocated 82MB on [0] TITAN X (Pascal), 11939MB remaining.
[19:39:09] Allocated 24MB on [0] TITAN X (Pascal), 11913MB remaining.
[0]     test-error:0.056312
[19:39:09] Allocated 87MB on [0] TITAN X (Pascal), 11809MB remaining.
[1]     test-error:0.055784
[2]     test-error:0.055256
[3]     test-error:0.053432
[4]     test-error:0.053016
Train Time: 1.210120677947998 seconds
dt prepare2 Time: 0.00038504600524902344 seconds
dt->DMatrix Start
dt->DMatrix1 Time: 0.23475217819213867 seconds
dt->DMatrix2 Time: 0.07418322563171387 seconds
Training with 'gpu_hist'
[19:39:10] Allocated 82MB on [0] TITAN X (Pascal), 11939MB remaining.
[19:39:11] Allocated 24MB on [0] TITAN X (Pascal), 11913MB remaining.
[0]     test-error:0.056312
[19:39:11] Allocated 87MB on [0] TITAN X (Pascal), 11809MB remaining.
[1]     test-error:0.055784
[2]     test-error:0.055256
[3]     test-error:0.053432
[4]     test-error:0.053016
Train Time: 1.1787691116333008 seconds
```
This file and its output are pretty unintelligible to me, but it looks like the version that calls dtdata_X_train.tonumpy() might be the fastest on my system. Is this what is supposed to happen? If it is faster to convert the dt to np first then why do we need a custom DMatrix constructor?

Also I can't see any actual tests that would check for accuracy or correctness.

**Comment by pseudotensor** [comment]
Here's the slightly updated benchmark code so output is cleaner:
```
np->DMatrix Time(order=F): 0.21093034744262695 seconds
dt->DMatrix Time(order=F): 0.2873115539550781 seconds
dt->numpy->DMatrix Time(order=F): 0.355623722076416 seconds
np->DMatrix Time(order=C): 0.14051151275634766 seconds
dt->DMatrix Time(order=C): 0.2861809730529785 seconds
dt->numpy->DMatrix Time(order=C): 0.33972811698913574 seconds
```

**Comment by RAMitchell** [comment]
Ok thanks. Regarding tests here are some of the kinds of things you need to check in c++ and/or python:

Create a dmatrix from dt and check:
Number of rows/columns/nonzeroes are correct in dmatrix
Feature names are correct
Dt->Np->dmatrix is exactly same as dt->dmatrix

A couple of other issues:

Installing data table on Linux requires a direct link to an s3 hosted wheel. If this wheel changes all of our unit tests fail.

All printf type statements in your code need to be changed to use the xgboost LOG macros.

The dt.h header file is pretty nonstandard in xgb.  We don't use a lot of header files that are not interfaces. This code can be reduced by a lot and put in c_api.cc. I can help with this but I would be good to have the basic unit tests first so I can tell if I break it.

**Comment by pseudotensor** [comment]
Thanks.  On the installing of dt, You can just do pip install datatable.

File: include/xgboost/c_api.h [hunk 1]
```diff
@@ -261,7 +277,7 @@ XGB_DLL int XGDMatrixSetFloatInfo(DMatrixHandle handle,
  * \brief set uint32 vector to a content in info
  * \param handle a instance of data matrix
  * \param field field name
- * \param array pointer to float vector
+ * \param array pointer to pointer to columns
```
Comment (hcho3) [inline include/xgboost/c_api.h]: Shouldn't it say `array pointer to unsigned int vector`?

File: python-package/xgboost/core.py [hunk 2]
```diff
@@ -219,6 +218,55 @@ def _maybe_pandas_label(label):
...
+        raise ValueError(msg + ', '.join(bad_fields))
+
+    if feature_names is None:
+        feature_names = data.names
+
+        # always return stypes for dt ingestion
+        if feature_types is not None:
+            raise ValueError('DataTable has own feature types, cannot pass them in')
+        else:
+            data_types_names = tuple(lt.name for lt in data.ltypes)
```
Comment (hcho3) [inline python-package/xgboost/core.py]: This line appears to be redundant, as `data_types_names` is already computed in line 233.

File: python-package/xgboost/core.py [hunk 3]
```diff
@@ -219,6 +218,55 @@ def _maybe_pandas_label(label):
...
+
+    return data, feature_names, feature_types
+
+
+def _maybe_dt_array(array):
+    """ Extract numpy array from single column data table """
+    if not isinstance(array, DataTable) or array is None:
+        return array
+
+    if array is not None and array.shape[1] > 1:
```
Comment (hcho3) [inline python-package/xgboost/core.py]: Do we need to check `array != None` here? You are already checking it in line 257. Is this check for clarity's sake?

File: python-package/xgboost/core.py [hunk 4]
```diff
@@ -279,19 +333,23 @@ def __init__(self, data, label=None, missing=None,
...
+
         if label is not None:
-            if isinstance(data, np.ndarray):
+            if isinstance(label, np.ndarray):
                 self.set_label_npy2d(label)
             else:
                 self.set_label(label)
         if weight is not None:
-            if isinstance(data, np.ndarray):
+            if isinstance(weight, np.ndarray):
```
Comment (hcho3) [inline python-package/xgboost/core.py]: Nice catch

File: python-package/xgboost/core.py [hunk 5]
```diff
@@ -365,6 +423,33 @@ def _init_from_npy2d(self, mat, missing, nthread):
...
+        ptrs = (ctypes.c_void_p * data.ncols)()
+        for icol in range(data.ncols):
+            col = data.internal.column(icol)
+            cols.append(col)
+            # int64_t (void*)
+            ptr = col.data_pointer
+            ptrs[icol] = ctypes.c_void_p(ptr)
+
+        # always return stypes for dt ingestion
+        feature_stypes = (ctypes.c_wchar_p * data.ncols)()
```
Comment (hcho3) [inline python-package/xgboost/core.py]: What does `stypes` stand for? Does it stand for string types?

File: python-package/xgboost/core.py [hunk 6]
```diff
@@ -365,6 +423,33 @@ def _init_from_npy2d(self, mat, missing, nthread):
...
+        for icol in range(data.ncols):
+            feature_stypes[icol] = ctypes.c_wchar_p(data.stypes[icol].name)
```
Comment (hcho3) [inline python-package/xgboost/core.py]: Can we combine this loop with the loop above, since both loops loop from 0 to `data.ncols`.

Comment (hcho3) [reply python-package/xgboost/core.py]: Also, let us re-consider the choice of `c_wchar_p` (see my comment in `c_api.h`). I think we can use `c_char_p` here by using `string.encode` to obtain UTF-8 strings.

File: src/c_api/c_api.cc [hunk 7]
```diff
@@ -497,12 +499,139 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,  // NOLINT
       }
     }
   }
+  // restore omp state
+  omp_set_num_threads(nthread_orig);
```
Comment (hcho3) [inline src/c_api/c_api.cc]: Was there a particular issue that prompted you to save and restore omp states?

File: src/c_api/c_api.cc [hunk 8]
```diff
@@ -497,12 +499,139 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,  // NOLINT
...
 
   mat.info.num_nonzero_ = mat.page_.data.size();
   *out  = new std::shared_ptr<DMatrix>(DMatrix::Create(std::move(source)));
   API_END();
 }
 
+int DTGetType(const wchar_t* type_string) {
+  if (wcscmp(type_string, L"float32") == 0) {
```
Comment (hcho3) [inline src/c_api/c_api.cc]: You should include header `<cwchar>` to use `wcscmp`. Alternatively, use `strcmp` from `<cstring>`. (To choose this alternative, make sure to obtain UTF-8 encoded strings from the Python end.)

File: src/c_api/c_api.cc [hunk 9]
```diff
@@ -497,12 +499,139 @@ XGB_DLL int XGDMatrixCreateFromMat_omp(const bst_float* data,  // NOLINT
...
+    return 0;
```
Comment (hcho3) [inline src/c_api/c_api.cc]: I'm nervous about returning magic numbers here, as the return value of `DTGetType()` will be directly used by `DTGetValue()`. Can we create an enum class to create a group of integer constants? Something like
```cpp
enum class DTType : uint8_t {
  kFloat32 = 0,
  kFloat64 = 1,
  kBool8 = 2,
  kInt32 = 3,
  kInt8 = 4,
  kInt16 = 5,
  kInt64 = 6
};
```
Then the `DTGetType()` function will be like
```cpp
DTType DTGetType(const wchar_t* type_string) {
  if (wcscmp(type_string, L"float32") == 0) {
    return DTType::kFloat32;
  }  else if (wcscmp(type_string, L"float64") == 0) {
    return DTType::kFloat64;
  } /* ... other types ... */
}
```
and the `DTGetValue()` will be like
```cpp
float DTGetValue(void* column, DTType dt_type, size_t ridx) {
  float missing = std::numeric_limits<float>::quiet_NaN();
  switch (dt_type) {
     case DTType::kFloat32: { /* ... */ }
     case DTType::kFloat64: { /* ... */ }
     /* ... other types ... */
  }
}
```

File: tests/travis/run_test.sh [hunk 10]
```diff
@@ -48,6 +48,13 @@ if [ ${TASK} == "python_test" ]; then
     source activate python3
     python --version
     conda install numpy scipy pandas matplotlib nose scikit-learn
+
+    # Install data table from source
+    wget http://releases.llvm.org/5.0.2/clang+llvm-5.0.2-x86_64-linux-gnu-ubuntu-14.04.tar.xz
+    tar xf clang+llvm-5.0.2-x86_64-linux-gnu-ubuntu-14.04.tar.xz
+    export LLVM5=$(pwd)/clang+llvm-5.0.2-x86_64-linux-gnu-ubuntu-14.04
+    python -m pip install datatable
```
Comment (hcho3) [inline tests/travis/run_test.sh]: Make sure that `pip` use the source tarball, with command-line option `--no-binary datatable`.

**Comment by RAMitchell** [comment]
@hcho3 thanks, have addressed your comments.

**Comment by hcho3** [comment]
@pseudotensor @RAMitchell All look good to me. Will merge after tests pass.

**Pull request merged**